### PR TITLE
G2P

### PR DESCRIPTION
**Pull request by bwalsh**
This PR contains:
* G2P transformer
* Introduction of test helper to encapsulate basic vertex and edge tests
* Refactor to create bmeg.enrichers
* creates Deadletter vertex file to collect missing data
* creates MinimalAllele vertex for incomplete Alleles
* addresses #127

### CODE REVIEW

File: src/bmeg/vertex.py [hunk 1]
```diff
@@ -43,10 +43,10 @@ class Allele(Vertex):
     chromosome: str
     start: int
     end: int
-    reference_bases: str
-    alternate_bases: str
-    annotations: list = None
-    myvariantinfo: dict = None
+    reference_bases: Union[None, str]
```
Comment (adamstruck) [inline src/bmeg/vertex.py]: In what case would we want to allow reference_bases or alternate_bases to be None?

File: src/bmeg/edge.py [hunk 2]
```diff
@@ -199,3 +199,48 @@ class DrugResponseIn(Edge):
...
+class HasSupportingReference(Edge):
+    """
+    G2PAssociation -> Publication
+    """
+    pass
+
+
+@enforce_types
+@dataclass(frozen=True)
+class GeneFeatureFor(Edge):
```
Comment (adamstruck) [inline src/bmeg/edge.py]: To me this class name (and AlleleFeatureFor, PhenotypeOf, etc) make me expect the edge to point in the opposite direction.

Maybe HasGeneFeature?

File: transform/g2p/transform.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        yield NormalizedAssociation(hit, genes, features, environments,
+                                    phenotypes, publications, association)
+
+
+def toGraph(normalized_association, emitter):
+    """ tuple to graph edges and vertexes """
+    na = normalized_association
+    association = na.association
+    emitter.emit_vertex(association)
+    for publication in na.vertices['publications']:
```
Comment (adamstruck) [inline transform/g2p/transform.py]: I think we should leave it to the pubmed transformer to create the publication vertices.

File: src/bmeg/vertex.py (hunk 1 above)
Comment (bwalsh) [reply src/bmeg/vertex.py]: g2p `features` missing ref | alt

without union for ref, alt:
```
from_source	159
is_amplification	6515
is_deletion	3128
is_fusion_acceptor	4620
is_fusion_donor	4463
is_inact_mut	197
is_loss	6895
is_mutation	2783
is_missing_location	13205
```

with Union[None,str] for ref, alt:

```
is_missing_location	13205
```

File: transform/g2p/transform.py (hunk 3 above)
Comment (bwalsh) [reply transform/g2p/transform.py]: agreed

File: src/bmeg/vertex.py (hunk 1 above)
Comment (adamstruck) [reply src/bmeg/vertex.py]: Amplification / deletion I would think refers to a copy number alteration not an allele. The same goes for fusions - they are structural events not alleles. @kellrott what do you think?

**Comment by bwalsh** [comment]
@adamstruck: made suggested changes.  Can you take another look?

File: src/bmeg/util/logging.py [hunk 4]
```diff
@@ -1,16 +1,16 @@
 
 """ utility, minimal logging   """
 import logging
-import json
+# import json
```
Comment (adamstruck) [inline src/bmeg/util/logging.py]: Please remove unused code.

File: src/bmeg/util/logging.py [hunk 5]
```diff
@@ -1,16 +1,16 @@
...
     """ bare bones logging"""
     fmt = '%(asctime)s - %(levelname)s - %(threadName)s - %(message)s'
     logging.basicConfig(level=loglevel, format=fmt)
-
-
-def log_missing_vertex(msg):
-    """ standard way to log missing data"""
-    msg['DATA_EXCEPTION'] = 'MISSING_VERTEX'
-    logging.warning(json.dumps(msg, separators=(',', ':')))
+#
```
Comment (adamstruck) [inline src/bmeg/util/logging.py]: Please remove unused code.

File: src/bmeg/vertex.py [hunk 6]
```diff
@@ -43,10 +43,12 @@ class Allele(Vertex):
     chromosome: str
     start: int
     end: int
-    reference_bases: Union[None, str]
-    alternate_bases: Union[None, str]
+    reference_bases: str
+    alternate_bases: str
     annotations: Union[None, list] = None
     myvariantinfo: Union[None, dict] = None
+    # type: Union[None, str] # SNP, ...
```
Comment (adamstruck) [inline src/bmeg/vertex.py]: Please remove unused code.

File: transform/g2p/environments.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+
+from bmeg.vertex import Compound
+
+# keep track of what we've already exported
+EXPORTED_COMPOUNDS = []
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply transform/g2p/environments.py]: @buchanae see lines 20, 21 below

[the remark this answers is not in the document]
Comment (buchanae) [reply transform/g2p/environments.py]: oh, duh. my bad.

**Comment by buchanae** [comment]
Other than Travis, it looks good to me. Hope we can get this merged today/tomorrow. I'd like to use some of this in my gene expression branch.

**Pull request merged**